usr/src/uts/common/io/overlay/overlay_router.c

```diff
@@ -0,0 +1,2389 @@
+	mutex_exit(&rtab->ort_lock);
+
+	if (cnt != 0)
+		return;
+
+	VERIFY(!list_link_active(&rtab->ort_link));
+
+	if (rtab->ort_routev4 != NULL) {
+		kmem_free(rtab->ort_routev4,
+		    rtab->ort_routev4_alloc * sizeof (overlay_route_ent_t));
+		rtab->ort_routev4 = NULL;
+	}
+
+	if (rtab->ort_routev6 != NULL) {
+		kmem_free(rtab->ort_routev6,
+		    rtab->ort_routev6_alloc * sizeof (overlay_route_ent_t));
+		rtab->ort_routev6 = NULL;
+	}
+
+	kmem_cache_free(overlay_rtab_cache, rtab);
+}
+
+static overlay_net_t *
+overlay_hold_net_by_id(overlay_router_t *orr, const char *id)
+{
+	overlay_net_t *ont;
+
+	mutex_enter(&orr->orr_lock);
+	/*
+	 * Since every net must have a valid MAC address, we just
+	 * search the mac AVL tree. We can always do something different
+	 * in the future if this turns out to be too slow.
+	 */
+	for (ont = avl_first(&orr->orr_nets_mac); ont != NULL;
+	    ont = AVL_NEXT(&orr->orr_nets_mac, ont)) {
+		if (strcmp(ont->ont_id, id) == 0) {
+			mutex_enter(&ont->ont_lock);
+			ont->ont_refcnt++;
+			mutex_exit(&ont->ont_lock);
+			mutex_exit(&orr->orr_lock);
+			return (ont);
+		}
+	}
+	mutex_exit(&orr->orr_lock);
+	return (NULL);
+}
+
+/* vlan is in host byte order */
+overlay_net_t *
+overlay_hold_net_by_vlan(overlay_router_t *orr, uint16_t vlan)
+{
+	overlay_net_t *ont = NULL;
+	overlay_net_t ref = {
+		.ont_vlan = vlan
+	};
+
+	mutex_enter(&orr->orr_lock);
+	ont = avl_find(&orr->orr_nets_vlan, &ref, NULL);
+	if (ont != NULL) {
+		mutex_enter(&ont->ont_lock);
+		ont->ont_refcnt++;
+		mutex_exit(&ont->ont_lock);
+	}
+	mutex_exit(&orr->orr_lock);
+	return (ont);
+}
+
+overlay_net_t *
+overlay_hold_net_by_mac(overlay_router_t *orr, const uint8_t mac[ETHERADDRL])
+{
+	overlay_net_t *ont = NULL;
+	overlay_net_t ref = { 0 };
+
+	bcopy(mac, ref.ont_mac, ETHERADDRL);
+
+	mutex_enter(&orr->orr_lock);
+	ont = avl_find(&orr->orr_nets_mac, &ref, NULL);
+	if (ont != NULL) {
+		mutex_enter(&ont->ont_lock);
+		ont->ont_refcnt++;
+		mutex_exit(&ont->ont_lock);
+	}
+	mutex_exit(&orr->orr_lock);
+	return (ont);
+}
+
+/*
+ * Find and hold the overlay_net_t whose subnet contains addr, and return the
+ * overlay_net_t, or NULL if not found.
+ */
+overlay_net_t *
+overlay_hold_net_by_ip(overlay_router_t *orr, in_addr_t addr)
+{
+	overlay_net_t *ont;
+	avl_index_t where;
+	overlay_net_t ref = {
+		.ont_net = addr
+	};
+
+	mutex_enter(&orr->orr_lock);
+
+	/*
+	 * The nets_v4 avl tree orders the entries by the network address,
+	 * e.g. 10.0.1.0, 172.16.5.0, 192.168.1.0, 200.200.0.0, ...
+	 *
+	 * We use the fact that the network address is always the first
+	 * (and thus lowest numerically) address in a subnet. When we search
+	 * for the address, it should normally fail (unless it's an actual
+	 * network address we have), however if the network exists, the
+	 * AVL_BEFORE entry should be the network address that contains
+	 * 'addr'. We verify and return, otherwise the network isn't present.
+	 */
+	ont = avl_find(&orr->orr_nets_v4, &ref, &where);
```

Comment:
Yeah, I tried to isolate that so it would hopefully be easier to swap out with a different method in the future -- I figured early on AVL trees should be good enough (and was something already there), but it would be worth seeing if other storage methods (like a radix tree or a variant like poptrie) might be useful.
It might be worth adding checks for overlapping networks and disallow. But since initially, NAPI is doing the same checks, I didn't duplicate it here (though I could add a comment for future work).

Reply:
Yeah, it's a science experiment, not something to block you right now.